hal/src/common/thumbv7em/clock/v2/sources/dfll.rs

```diff
@@ -1,103 +1,344 @@
     }
 }
 
-impl<N: Count> Sealed for Dfll<N> {}
-
+impl<TMode: Mode, N: Count> Sealed for Dfll<TMode, N> {}
 
 //==============================================================================
 // Lockable
 //==============================================================================
 
-impl<N> Lockable for Dfll<N>
+impl<TMode, N> Lockable for Dfll<TMode, N>
 where
+    TMode: Mode,
     N: Increment,
 {
-    type Locked = Dfll<N::Inc>;
+    type Locked = Dfll<TMode, N::Inc>;
     fn lock(self) -> Self::Locked {
-        let Dfll { regs, count } = self;
+        let Dfll { count, config } = self;
         let count = count.inc();
-        Dfll { regs, count }
+        Dfll { count, config }
     }
 }
 
 //==============================================================================
 // Unlockable
 //==============================================================================
 
-impl<N> Unlockable for Dfll<N>
+impl<TMode, N> Unlockable for Dfll<TMode, N>
 where
+    TMode: Mode,
     N: Decrement,
 {
-    type Unlocked = Dfll<N::Dec>;
+    type Unlocked = Dfll<TMode, N::Dec>;
     fn unlock(self) -> Self::Unlocked {
-        let Dfll { regs, count } = self;
+        let Dfll { count, config } = self;
         let count = count.dec();
-        Dfll { regs, count }
+        Dfll { count, config }
     }
 }
 
 //==============================================================================
 // GclkSource
 //==============================================================================
 
-pub enum Fll {}
-
-impl Sealed for Fll {}
-
-impl GclkSourceType for Fll {
-    const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::DFLL;
+impl<G: GenNum, N: Count> GclkSource<G> for Dfll<OpenMode, N> {
+    type Type = marker::Dfll<OpenMode>;
+    #[inline]
+    fn freq(&self) -> Hertz {
+        self.freq()
+    }
 }
 
-impl<G: GenNum, N: Count> GclkSource<G> for Dfll<N> {
-    type Type = Fll;
+impl<G: GenNum, T: PclkSourceType, N: Count> GclkSource<G> for Dfll<ClosedMode<T>, N> {
+    type Type = marker::Dfll<marker::ClosedMode>;
     #[inline]
     fn freq(&self) -> Hertz {
         self.freq()
     }
 }
+
+pub mod marker {
+    use super::{GclkSourceEnum, GclkSourceType, Sealed};
+    use core::marker::PhantomData;
+
+    pub trait ModeMarker: Sealed {}
+    /// TODO
+    /// super::ClosedMode type is polluted with a generic parameter describing
+    /// reference clock. It is undesirable to have a marker type owned by a
+    /// Gclk that knows about source of its source.
+    /// This is a reason for existence of this type.
+    pub struct ClosedMode {
+        __: (),
+    }
+
+    impl Sealed for ClosedMode {}
+    impl ModeMarker for ClosedMode {}
+
+    impl ModeMarker for super::OpenMode {}
+
+    pub struct Dfll<T: ModeMarker> {
+        __: PhantomData<T>,
+    }
+
+    impl<T: ModeMarker> Sealed for Dfll<T> {}
+
+    impl<T: ModeMarker> GclkSourceType for Dfll<T> {
+        const GCLK_SRC: GclkSourceEnum = GclkSourceEnum::DFLL;
+    }
+}
```

Comment:
I realized recently that we probably shouldn't reuse existing types for the implementers of `GclkSourceType`. It's too confusing to determine which type to use, and we end up creating new types with nearly identical names.

I think we should create a dedicated type-level enum. Basically just lift `GclkSourceEnum` to the type level, with dedicated types. They can be called `Dpll0Source`, `Gclk1Source`, etc. (or similar). Basically, add a consistent prefix or suffix to the type to distinguish it as a `GclkSourceType`.

Also, there's no need for this marker approach. There can only ever be one `Dfll`. A `Gclk` does not need to know or care what `Mode` it's in. We just need a single `DfllSource` to implement `GclkSourceType`.

Reply:
> Also, there's no need for this marker approach. 

I dislike the fact that now `GClk` knows about the mode `Dfll` is using but I needed it to create impl specializations for `change_mode` methods. I'll tag you in a relevant spot so you can take a look.

I agree that reusing existing types is quite messy. For protytyping sake I just created a marker type for `ClosedMode` because in contrast to `super::OpenMode`, `super::ClosedMode<T>` is polluted with a generic type.